konnectivity-socks5-proxy/main.go

```diff
@@ -194,18 +225,24 @@ func (gr *guestClusterResolver) resolve(ctx context.Context, name string) (net.I
 }
 
 // proxyResolver tries to resolve addresses using the following steps in order:
-// 1. Not at all for cloud provider apis, as we do not want to tunnel them through Konnektivity
-// 2. If the address is a valid Kubernetes service and that service exists in the guest cluster, it's clusterIP is returned
-// 2. If --resolve-from-guest-cluster-dns is set, it uses the guest clusters dns. If that fails, an error is returned
-// 4. Lastly, golangs default resolver is used
+// 1. Not at all for cloud provider apis, as we do not want to tunnel them through Konnectivity.
+// 2. If the address is a valid Kubernetes service and that service exists in the guest cluster, it's clusterIP is returned.
+// 3. If --resolve-from-guest-cluster-dns is set, it uses the guest clusters dns. If that fails, fallback to the management cluster's resolution.
+// 4. Lastly, Golang's default resolver is used.
 type proxyResolver struct {
 	client                  client.Client
 	resolveFromGuestCluster bool
+	dnsFallbackRWMutex      *sync.RWMutex
+	dnsFallback             *bool
 	guestClusterResolver    *guestClusterResolver
 	log                     logr.Logger
 }
 
 func (d proxyResolver) Resolve(ctx context.Context, name string) (context.Context, net.IP, error) {
+	d.dnsFallbackRWMutex.Lock()
+	*d.dnsFallback = false
```

Comment:
This may not be a good thing to do because you may have another async dial request that will fail because it thinks it should not fallback. 

The behavior I'd think makes sense for this variable is:
- If unset, assume it is false
- Set to true only if unset
- Set to false when resolution works once. Never set back to false.